core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java

```diff
@@ -195,22 +210,29 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
             for (CatalogItem<?, ?> item : items) {
                 LOG.debug("Unloading {} {} from catalog", item.getSymbolicName(), item.getVersion());
 
-                removeFromCatalog(item);
+                catalogBundleLoader.removeFromCatalog(item);
             }
         }
+    }
 
-        private void removeFromCatalog(CatalogItem<?, ?> item) {
-            try {
-                getManagementContext().getCatalog().deleteCatalogItem(item.getSymbolicName(), item.getVersion());
-            } catch (Exception e) {
-                Exceptions.propagateIfFatal(e);
-                LOG.warn(Strings.join(new String[] {
-                    "Failed to remove", item.getSymbolicName(), item.getVersion(), "from catalog"
-                }, " "), e);
-            }
+    @Beta
+    public class CatalogBundleLoader {
```

Comment:
I'd say it's worth splitting this out into a standalone class. Same for CatalogPopulator; which I think we should rename to something like 'CatalogLoadingBundleTracker'.